fqn_decorators/decorators.py

```diff
@@ -36,10 +36,13 @@ class Decorator(object):
     A base class to easily create decorators.
     """
 
-    def __init__(self, func=None, **params):
+    def __init__(self, func=None, _initialized=False, **params):
```

Comment:
why do we need this flag to be passed in?

Reply:
When decorator is applied (without arguments for example)
```python
@decorator
def fun(a):
    pass
```
Then its `init` method is called immediately upon function initialization:
```
fun = decorator(fun)
```
And when function is actually called, we dive into `__call__` method of already created decorator object. And that was the issue: all function calls share same decorated function object. So as a fix here I force to create another decorator instance upon each `__call__` just for this very function call. But it means we introduce recursion here:
* `decorator.__init__` -> 
* `decorator.__call__` -> 
* let's create a new object then -> 
* `decorator.__init__` -> 
* `decorator.__call__` -> 
* ok, let's create a new object -> ...     # we want to prevent this

To prevent going into next level of recursion, I use a flag upon decorator creation to show that we are already in a separate instance for the call and there's no need to go deeper.